asm-to-pil/src/vm_to_constrained.rs

```diff
@@ -454,6 +456,86 @@ impl<T: FieldElement> ASMPILConverter<T> {
         res
     }
 
+    fn handle_instr_declaration_params(&mut self, m: &mut Machine<T>) {
+        for l in &mut m.links {
+            let lhs = &l.params;
+            let rhs = &mut l.to.params;
+            // lhs params must all be assignment registers when mapping instruction to operation
+            lhs.inputs.iter().chain(lhs.outputs.iter()).for_each(|p| {
+                assert!(p.ty.is_none());
+                assert!(self
+                    .registers
+                    .get(&p.name)
+                    .is_some_and(|r| r.ty.is_assignment()));
+            });
+
+            // if rhs is not empty, check it's valid
+            if !rhs.is_empty() {
+                // rhs params must either be assignment registers declared in the lhs or write registers
+                rhs
+                .inputs
+                .iter_mut()
+                .chain(rhs.outputs.iter_mut())
+                .for_each(|p| {
+                    assert!(p.ty.is_none());
+                    let reg = self.registers.get(&p.name).expect("rhs params must be registers");
+                    match reg.ty {
+                        RegisterTy::Assignment => {
+                            let mut lhs_params = lhs.inputs.iter().chain(lhs.outputs.iter());
+                            assert!(lhs_params.any(|p_lhs| p_lhs.name == p.name),
+                                    "assignment register on rhs must be present on lhs params");
+                        }
+                        RegisterTy::Write => {
+                            // the linker needs this information to build the
+                            // plookup when a write register is used as output.
+                            // TODO: less hackish way of passing this info? proper type for param.ty?
```

Comment:
yea agree

Reply:
I'll do this change on a follow-up PR as it touches some other unrelated stuff.

Reply:
can you make an issue so we don't forget